crates/valence_advancement/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+) {
+    for advancement_cached_bytes in cache_clear_query.iter() {
+        advancement_cached_bytes.0.lock().clear()
+    }
+
+    for (mut advancement_client_update, mut client) in client.iter_mut() {
+        match advancement_client_update.force_tab_update {
+            ForceTabUpdate::None => {}
+            ForceTabUpdate::First => {
+                client.write_packet(&SelectAdvancementTabS2c { identifier: None })
+            }
+            ForceTabUpdate::Spec(spec) => {
+                if let Ok(a_identifier) = update_single_query.advancement_id_query.get(spec) {
+                    client.write_packet(&SelectAdvancementTabS2c {
+                        identifier: Some(a_identifier.0.borrowed()),
+                    });
+                }
+            }
+        }
+
+        if ForceTabUpdate::None != advancement_client_update.force_tab_update {
+            advancement_client_update.force_tab_update = ForceTabUpdate::None;
+        }
+
+        if advancement_client_update.new_advancements.is_empty()
+            && advancement_client_update.progress.is_empty()
+            && advancement_client_update.remove_advancements.is_empty()
+        {
+            continue;
+        }
+
+        let advancement_client_update = std::mem::take(advancement_client_update.as_mut());
+
+        client.write_packet(&AdvancementUpdateS2COE {
+            queries: &update_single_query,
+            client_update: advancement_client_update,
+        });
+    }
+}
+
+#[derive(Component)]
+pub struct Advancement(Ident<Cow<'static, str>>);
+
+impl Advancement {
+    pub fn new(ident: Ident<Cow<'static, str>>) -> Advancement {
+        Self(ident)
+    }
+
+    pub fn get(&self) -> &Ident<Cow<'static, str>> {
+        &self.0
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum AdvancementFrameType {
+    Task,
+    Challenge,
+    Goal,
+}
+
+#[derive(Component)]
+pub struct AdvancementDisplay {
+    pub title: Text,
+    pub description: Text,
+    pub icon: Option<ItemStack>,
+    pub frame_type: AdvancementFrameType,
+    pub show_toast: bool,
+    pub hidden: bool,
+    pub background_texture: Option<Ident<Cow<'static, str>>>,
+    pub x_coord: f32,
+    pub y_coord: f32,
+}
+
+impl AdvancementDisplay {
+    pub(crate) fn flags(&self) -> i32 {
+        let mut flags = 0;
+        flags |= self.background_texture.is_some() as i32;
+        flags |= (self.show_toast as i32) << 1;
+        flags |= (self.hidden as i32) << 2;
+        flags
+    }
+}
+
+#[derive(Component)]
+pub struct AdvancementCriteria(Ident<Cow<'static, str>>);
+
+impl AdvancementCriteria {
+    pub fn new(ident: Ident<Cow<'static, str>>) -> Self {
+        Self(ident)
+    }
+
+    pub fn get(&self) -> &Ident<Cow<'static, str>> {
+        &self.0
+    }
+}
+
+#[derive(Component)]
+pub struct AdvancementRequirements(pub Vec<Vec<Entity>>);
+
+#[derive(Component)]
+pub(crate) struct AdvancementCachedBytes(pub(crate) Mutex<Vec<u8>>);
```

Comment:
This mutex should be avoided somehow. Can you encode the necessary data into the cache before encoding the packet? That way, you won't need `&mut` when encoding the packet.